### PR TITLE
Add org.openmw.OpenMW

### PR DESCRIPTION
**Pull request by ace13**
The unofficial open source engine reimplementation of the game Morrowind.

Plenty of large dependencies, Qt4 also required the OpenGL module which seems to be disabled on the shared module.

### CODE REVIEW

File: org.openmw.OpenMW.json [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+          "libQtCLucene*",
+          "libQtDesigner*",
+          "libQtHelp*",
+          "libQtSql*",
+          "libQtXml*"
+        ]
+      },
+      {
+        "name": "openscenegraph",
+        "buildsystem": "cmake",
```
Comment (nedrichards) [inline org.openmw.OpenMW.json]: If this works with cmake-ninja then it'd be a lot faster.

File: org.openmw.OpenMW.json [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+        "cleanup": [
+          "/include",
+          "/lib/*.a",
+          "/lib/*.la",
+          "/lib/pkgconfig"
+        ]
+      },
+      {
+        "name": "libunshield",
+        "buildsystem": "cmake",
```
Comment (nedrichards) [inline org.openmw.OpenMW.json]: As above on cmake-ninja

File: org.openmw.OpenMW.json [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+          "libboost_type_erasure*",
+          "libboost_unit_test_framework*",
+          "libboost_wave*",
+          "libboost_wserialization*"
+        ]
+      },
+      {
+        "name": "openmw",
+        "builddir": true,
+        "buildsystem": "cmake",
```
Comment (nedrichards) [inline org.openmw.OpenMW.json]: As above on cmake-ninja

File: org.openmw.OpenMW.json [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+              "sed -i s:Icon=openmw:Icon=org.openmw.OpenMW.png: files/openmw.desktop",
+              "sed -i s:Icon=openmw-cs:Icon=org.openmw.OpenMW.OpenCS.png: files/openmw-cs.desktop"
+            ]
+          }
+        ],
+        "post-install": [
+          "mv /app/share/applications/openmw.desktop /app/share/applications/org.openmw.OpenMW.desktop",
+          "mv /app/share/applications/openmw-cs.desktop /app/share/applications/org.openmw.OpenMW.OpenCS.desktop",
+
+          "mkdir -p /app/share/appdata",
```
Comment (nedrichards) [inline org.openmw.OpenMW.json]: is there a reason why you don't use the builtin 'rename-appdata' 'rename-icon' and 'rename-desktop-file' stuff?

**Comment by nedrichards** [comment]
Test build https://flathub.org/builds/#/builders/1/builds/799

File: org.openmw.OpenMW.json [hunk 5]
```diff
@@ -0,0 +1,360 @@
+{
+  "app-id": "org.openmw.OpenMW",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "3.24",
```
Comment (nedrichards) [inline org.openmw.OpenMW.json]: not a blocker, but does this work on the more recent 3.26 runtime or are there dependencies that lock it here?

File: org.openmw.OpenMW.json [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+    "cxxflags": "-O2 -g"
+  },
+  "cleanup": ["/include", "/lib/pkgconfig",
+    "/share/pkgconfig", "/share/aclocal",
+    "/man", "/share/man", "/share/gtk-doc",
+    "/share/vala",
+    "*.la", "*.a"],
+    "modules": [
+      {
+        "name": "libmp3lame",
```
Comment (nedrichards) [inline org.openmw.OpenMW.json]: should use lame [from shared-modules](https://github.com/flathub/shared-modules/blob/master/lame/lame-3.99.5.json).

File: org.openmw.OpenMW.json [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        "config-opts": [
+          "-DBUILD_BSATOOL=no",
+          "-DBUILD_ESMTOOL=no",
+          "-DBUILD_MYGUI_PLUGIN=no"
+        ],
+        "sources" : [
+          {
+            "type": "git",
+            "url": "git://github.com/OpenMW/openmw.git",
+            "branch": "openmw-0.42.0"
```
Comment (nedrichards) [inline org.openmw.OpenMW.json]: Can this get a commit or a tag as well to specify the release more tightly. 

It'll be faster to build if you just switch it over to the source code archive [published here](https://github.com/OpenMW/openmw/releases).

**Comment by nedrichards** [comment]
Fails in lame on i386, if you switch to the shared-modules version it should work.

**Comment by nedrichards** [comment]
New build: https://flathub.org/builds/#/builders/1/builds/805

**Comment by ace13** [comment]
Wasn't able to get bullet to link properly when using ninja (Had honestly forgotten that existed), but changed all the other ones to it.  
Also couldn't get it to work with rename-icon, though desktop and appdata does seem to work with that.

Could probably build with a source package, but the build uses git data to generate the displayed version.

**Comment by nedrichards** [comment]
All of that sounds totally reasonable.

New build: https://flathub.org/builds/#/builders/1/builds/811

**Comment by ace13** [comment]
I get the feeling that maybe having a pre-build step on the buildbot that checks the manifest for validity (No unknown keys, no invalid values, etc) would save on build time.

First a wrongly named key `rename-appdata` vs `rename-appdata-file`, which should've probably thrown an error.  
And then I had the branch set to `master` from when I was writing this manifest, which should've also probably stopped the build before running through it and then failing on a validation stage.

Hopefully that'll be the last of the small mistakes though.

File: org.openmw.OpenMW.json [hunk 8]
```diff
@@ -0,0 +1,265 @@
+{
+  "app-id": "org.openmw.OpenMW",
+  "runtime": "org.gnome.Platform",
```
Comment (TingPing) [inline org.openmw.OpenMW.json]: If this uses Qt you probably don't need anything from this? Just use `org.freedesktop.Platform` if so.

File: org.openmw.OpenMW.json [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    "/share/doc",
+    "/man", "/share/man", "/share/gtk-doc",
+    "/share/vala",
+    "*.la", "*.a"],
+    "modules": [
+      "shared-modules/lame/lame-3.99.5.json",
+      {
+        "name": "ffmpeg",
+        "config-opts": [
+          "--enable-gpl",
```
Comment (TingPing) [inline org.openmw.OpenMW.json]: I imagine this only needs one or two codecs and can be made more minimal than this.

File: org.openmw.OpenMW.json [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+          "--disable-programs",
+          "--disable-doc",
+          "--disable-avdevice",
+          "--disable-avfilter",
+          "--disable-postproc"
+        ],
+        "sources": [
+          {
+            "type": "archive",
+            "url": "https://ffmpeg.org/releases/ffmpeg-3.2.4.tar.xz",
```
Comment (TingPing) [inline org.openmw.OpenMW.json]: Outdated

File: org.openmw.OpenMW.json [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+            "sha256": "6e38ff14f080c98b58cf5967573501b8cb586e3a173b591f3807d8f0660daf7a"
+          }
+        ]
+      },
+      {
+        "name": "qt4",
```
Comment (TingPing) [inline org.openmw.OpenMW.json]: Qt4 is in `shared-modules`.

File: org.openmw.OpenMW.json [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+            ],
+            "dest-filename": "configure"
+          }
+        ]
+      },
+      {
+        "name": "openmw",
+        "builddir": true,
+        "buildsystem": "cmake-ninja",
+        "config-opts": [
```
Comment (TingPing) [inline org.openmw.OpenMW.json]: Should pass `-DCMAKE_BUILD_TYPE=Release` to every CMake project.

File: org.openmw.OpenMW.json (hunk 11 above)
Comment (ace13) [reply org.openmw.OpenMW.json]: This requires the QtOpengl module, which is explicitly disabled in the shared module, so should it be activated there or done separately here?

Comment (TingPing) [reply org.openmw.OpenMW.json]: Hmm, it is fine to keep it here.

**Comment by nedrichards** [comment]
Build: https://flathub.org/builds/#/builders/1/builds/816

**Comment by TingPing** [comment]
x86_64 rebuild because of random git failure: https://flathub.org/builds/#/builders/4/builds/807

**Comment by ace13** [comment]
Is the git failure due to the use of git sources in the build?

I could change the manifest to grab sources from an archive if that's the case, inject revision data back into the build afterwards somehow.

File: org.openmw.OpenMW.json [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+              "./bootstrap.sh \"$@\"",
+              "cat <<EOF >Makefile",
+              "all:",
+              "\t./b2 -j `nproc` ./b2 stage threading=multi link=shared",
+              "",
+              "install:",
+              "\t./b2 install",
+              "EOF"
+            ],
+            "dest-filename": "configure"
```
Comment (Barthalion) [inline org.openmw.OpenMW.json]: Just use the `simple` buildsystem. if you need an example, look at `net.minetest.Minetest`. (Also some apps like `org.gnode.NixView` already use boost so you can copy it from there.)

File: org.openmw.OpenMW.json [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+        "sources" : [
+          {
+            "type": "git",
+            "url": "https://github.com/OpenMW/openmw.git",
+            "tag": "openmw-0.42.0"
+          },
+          {
+            "type": "shell",
+            "commands": [
+              "sed -i s:Icon=openmw:Icon=org.openmw.OpenMW.png: files/openmw.desktop",
```
Comment (Barthalion) [inline org.openmw.OpenMW.json]: Have you asked upstream if they could do that change themselves in the next release?

File: org.openmw.OpenMW.json [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+          "--disable-postproc",
+
+          /* Enough to run the base game, might need extending for mod content */
+          "--enable-decoder=bink,binkaudio_rdft,mp3,pcm_s16le,vorbis",
+          "--enable-demuxer=bink,mp3,ogg,pcm_s16le,wav"
+        ],
+        "sources": [
+          {
+            "type": "archive",
+            "url": "http://ffmpeg.org/releases/ffmpeg-3.4.tar.xz",
```
Comment (Barthalion) [inline org.openmw.OpenMW.json]: I'd rather use 3.3.x, at least in Arch we're having some issues with 3.4 across packages.

File: org.openmw.OpenMW.json [hunk 16]
```diff
@@ -0,0 +1,272 @@
+{
+  "app-id": "org.openmw.OpenMW",
+  "runtime": "org.freedesktop.Platform",
+  "runtime-version": "1.6",
+  "sdk": "org.freedesktop.Sdk",
+  "command": "openmw-launcher",
+  "rename-appdata-file": "openmw.appdata.xml",
+  "rename-desktop-file": "openmw.desktop",
+  "finish-args": [
+    "--share=ipc", "--socket=x11", "--device=dri",
```
Comment (Barthalion) [inline org.openmw.OpenMW.json]: I suppose that it needs full filesystem access to select Morrowind files?

File: org.openmw.OpenMW.json (hunk 14 above)
Comment (ace13) [reply org.openmw.OpenMW.json]: 0.43.0 is already in the late stage of the release phase - RCs have been done and just awaiting a release video, but changing the icon name might be possible to do for 0.44.0.

File: org.openmw.OpenMW.json (hunk 16 above)
Comment (ace13) [reply org.openmw.OpenMW.json]: If the user is planning on installing Morrowind from original discs, then the application won't need to access anything.

For those with local installs already though, my thought was to mention the use of something like `flatpak override --filesystem=~/game/morrowind:ro org.openmw.OpenMW` in the readme - and maybe even add an entry about it on the installation wizard when built as a flatpak package.  
But yeah, maybe just giving access to the filesystem would work okay too.

Comment (Barthalion) [reply org.openmw.OpenMW.json]: So I might be misunderstanding something, but with how the things are now, files need to be accessible somewhere at $HOME, as mount points aren't exposed unless `--filesystem=host` is passed.

File: org.openmw.OpenMW.json (hunk 14 above)
Comment (Barthalion) [reply org.openmw.OpenMW.json]: :+1:

**Comment by Barthalion** [comment]
bot, build org.openmw.OpenMW

File: org.openmw.OpenMW.json (hunk 16 above)
Comment (Barthalion) [reply org.openmw.OpenMW.json]: This is not a problem per se, but less technical users might be unhappy about having to fiddle with this from terminal (and there is no support for modifying permissions in GNOME Software yet). I'd pass `host` for now and if possible, add notice in the wizard you mentioned in future releases.

**Comment by nedrichards** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/org.openmw.OpenMW <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Pull request closed without merging**